WooCommerce/Classes/ViewRelated/Products/Media/DefaultProductImagesProvider.swift

```diff
@@ -0,0 +1,55 @@
+import Photos
+import Yosemite
+
+/// Provides images based on Product image status:
+/// - Requests the image from `PHImageManager` for a `PHASset`
+/// - Downloads the image given the URL of a remote Product image
+/// - Caches the images locally for remote Product images
+/// - When updating from an asset to remote Product image, caches the asset image locally to avoid an extra API request
+///
+final class DefaultProductImagesProvider: ProductImagesProvider {
+    private var imagesByProductImageID: [Int64: UIImage] = [:]
```

Comment:
`DefaultProductImagesProvider` is meant to provide images given any type of image resource in `ProductImageStatus`. Like in your second comment, `ImageService` caches the images that are downloaded from the Internet to the disk if requested. `ImageService` provides images for URLs, but not for `PHAsset`. I'm not sure if it makes sense for `ImageService` to provide images for `PHAsset` since `PHAsset` is most often used for image upload. what do you think?

Reply:
>  I'm not sure if it makes sense for ImageService to provide images for PHAsset

I'm not sure either. This is fine. 🙂 

---

So, my source of confusion was the usage of `downloadImage` in here:

https://github.com/woocommerce/woocommerce-ios/blob/f574d0aa8cd26708c6f4b0a56b3fc515c06fc958/WooCommerce/Classes/ViewRelated/Products/Media/DefaultProductImagesProvider.swift#L39-L42

I expected that `imageService.downloadImage` operates like this:

```
if imageIsInCache(url)
    return cachedImage
else 
   downloadImage(url)
   cacheTheImage()
   return downloadedImage
```

That's the standard seamless operation of image caching that I was used to. Every next call of `downloadImage` would trigger a cache check. 

But, now that I looked at [`DefaultImageService`](https://github.com/woocommerce/woocommerce-ios/blob/issue/1713-image-upload-observations/WooCommerce/Classes/Tools/ImageService/DefaultImageService.swift), I realized it doesn't work like that. 😄 

With this, I now wonder if we should move the caching from `DefaultProductImagesProvider` to `ImageService`. I think it'll make this simpler. Something to think about in the future! 

Sorry for the confusion. 🙂

Reply:
great question 💭 - I originally implemented `ImageService` to always download the image given a URL manually because the image URL didn't seem unique if the user replaces or deletes then uploads an edited version of the image. For example, the image URL from an image upload on a simulator looks like `{{site_url}}/wp-content/uploads/2020/02/img_0002-6.jpeg`.

after some testing, here are my observations on my AT site:
- Editing an image in core (Media tab in wp-admin): a new image URL is generated for the media resource - this is good so the client will download the new image
- Deleting an image in core, then uploading an edited version of the original image (like in the Photo Library app): the image URL will be the same as the original one. I tested by applying a warm filter to the image. It's interesting that core is showing the unedited version of the image, while the Product page shows the edited version:

WP Media library | Product page that has the image
-- | --
`{{site_url}}/wp-content/uploads/2020/02/f2ca4b0e-7422-4cd4-ac35-57c7c0c520b8-1041-0000007581a63da9_file-1.jpeg` | `https://i1.wp.com/{{site_url}}/wp-content/uploads/2020/02/f2ca4b0e-7422-4cd4-ac35-57c7c0c520b8-1041-0000007581a63da9_file-1.jpeg?fit=2250%2C3000&ssl=1`
<img width="516" alt="Screen Shot 2020-02-28 at 11 07 04 AM" src="https://user-images.githubusercontent.com/1945542/75507142-2fc0c200-5a1b-11ea-86c0-3b9806c86154.png"> | <img width="516" alt="Screen Shot 2020-02-28 at 11 12 05 AM" src="https://user-images.githubusercontent.com/1945542/75507155-351e0c80-5a1b-11ea-8577-4569f9ff3eb0.png">

I don't know too much about how media are stored and served for WP sites (maybe it also differs from the config), but I think it's pretty safe to retrieve an image from local cache like you mentioned:

```
if imageIsInCache(url)
    return cachedImage
else 
   downloadImage(url)
   cacheTheImage()
   return downloadedImage
```

I believe this is also the default behavior when retrieving an image in the third-party image library we use, [`Kingfisher`](https://github.com/onevcat/Kingfisher) - from the [suggestion on their wiki](https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet#getting-an-image-without-ui), it is the most common to call

```
KingfisherManager.shared.retrieveImage(with: url) { result in 
    // Do something with `result`
}
```

based on the [function-level comment](https://github.com/onevcat/Kingfisher/blob/master/Sources/General/KingfisherManager.swift#L141-L144):

> /// - Note:
    ///    This method will first check whether the requested `resource` is already in cache or not. If cached,
    ///    it returns `nil` and invoke the `completionHandler` after the cached image retrieved. Otherwise, it
    ///    will download the `resource`, store it in cache, then call `completionHandler`.

So I think we can update our current calls to download an image (`downloadImage(with url, ...)`) to like `retrieveImage(with url, ...)` similar to Kingfisher's image retrieval function above.

I created an issue https://github.com/woocommerce/woocommerce-ios/issues/1913 to improve `ImageService` implementation with more testing. Lemme know if you have any thoughts or concerns about using the `KingfisherManager.shared.retrieveImage` above in the `ImageService` implementation!

Reply:
> Deleting an image in core, then uploading an edited version of the original image (like in the Photo Library app): the image URL will be the same as the original one.

Ah, I can't say I haven't experienced this with other APIs before. 😒 Something to definitely test when we improve this.

Everything looks great! Thank you for the detailed response.